src/managedwriter/stream_connection.ts

```diff
@@ -106,44 +118,53 @@ export class StreamConnection extends EventEmitter {
 
   private handleError = (err: gax.GoogleError) => {
     this.trace('on error', err, JSON.stringify(err));
-    if (this.shouldReconnect(err)) {
-      this.reconnect();
-      return;
-    }
-    let nextPendingWrite = this.getNextPendingWrite();
     if (this.isPermanentError(err)) {
       this.trace('found permanent error', err);
-      while (nextPendingWrite) {
-        this.ackNextPendingWrite(err);
-        nextPendingWrite = this.getNextPendingWrite();
-      }
+      this.ackAllPendingWrites(err);
       this.emit('error', err);
       return;
     }
-    if (this.isRequestError(err) && nextPendingWrite) {
+    const nextPendingWrite = this.getNextPendingWrite();
+    if (nextPendingWrite) {
       this.trace(
         'found request error with pending write',
         err,
         nextPendingWrite
       );
-      this.ackNextPendingWrite(err);
+      const retrySettings = this._writeClient['_retrySettings'];
+      if (this.isRetryableError(err) && retrySettings.enableWriteRetries) {
+        process.nextTick(() => {
+          this.resendNextPendingWrite();
+        });
+      } else {
+        this.ackNextPendingWrite(err);
+      }
+      return;
+    }
+    if (this.isRetryableError(err) && this.listenerCount('error') === 0) {
       return;
     }
     this.emit('error', err);
```

Comment:
This line can be moved to L140? And the block of:
if (this.isPermanentError(err)) {
      this.trace('found permanent error', err);
      while (nextPendingWrite) {
        this.ackNextPendingWrite(err);
        nextPendingWrite = this.getNextPendingWrite();
      }
      this.ackAllPendingWrites(err);
      this.emit('error', err);
      return;
    }

Can be removed.

Reply:
what is the intention here ? nack each pending write individual in case of a permanent request error ? or just remove the concept of permanent errors ?

Reply:
If it is not retryable error then it is a permenant error. :-)

Reply:
resolve.